remote-drivers/lwip-ethernet-async/server.c

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright 2020, Data61
+ * Commonwealth Scientific and Industrial Research Organisation (CSIRO)
+ * ABN 41 687 119 230.
+ *
+ * This software may be distributed and modified according to the terms of
+ * the GNU General Public License version 2. Note that NO WARRANTY is provided.
+ * See "LICENSE_GPLv2.txt" for details.
+ *
+ * @TAG(DATA61_GPL)
+ */
+
+#include <autoconf.h>
+#include <stddef.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <camkes/dma.h>
+#include <camkes/dataport.h>
+#include <camkes/io.h>
+#include <camkes/irq.h>
+
+#include <platsupport/io.h>
+#include <platsupport/irq.h>
+#include <platsupport/interface_registration.h>
+#include <ethdrivers/raw.h>
+#include <ethdrivers/intel.h>
+#include <sel4utils/sel4_zf_logif.h>
+#include <lwip-ethernet-async.h>
+
+#include <utils/util.h>
+
+#include <shared_ringbuffer/shared_ringbuffer.h>
+#include <shared_ringbuffer/gen_config.h>
+
+#define BUF_SIZE 2048
+
+typedef struct data {
+    struct eth_driver *eth_driver;
+    /* mac address */
+    uint8_t hw_mac[6];
+    ps_io_ops_t *io_ops;
+
+    /* Pointers to shared_ringbuffers */
+    ring_handle_t *rx_ring;
+    ring_handle_t *tx_ring;
+} server_data_t;
+
+/* The camkes notification connection unregisters the callback each time. */
+static register_cb_fn reg_tx_cb;
+
+/**
+ * Packets have been transferred or dropped. Return the buffer
+ * into the available queue for reuse. 
+ * This is called by the actual driver.
+ *
+ * @param iface server data.
+ * @param cookie pointer to bufer descriptor to move.
+ *
+ */
+static void eth_tx_complete(void *iface, void *cookie)
+{   
+    ZF_LOGW("Packets have been transferred or dropped");
+    server_data_t *state = iface;
+
+    buff_desc_t *desc = cookie;
+
+    int err = enqueue_avail(state->tx_ring, desc->encoded_addr, desc->len, desc->cookie);
+    ZF_LOGF_IF(err, "lwip_eth_send: Error while enqueuing available buffer, tx available queue full");
+}
+
+/**
+ * Fetch an available buffer for receiving packets.   
+ * This is called by the actual driver.
+ *
+ * @param iface server data.
+ * @param buf_size size of buffer required. 
+ * @param cookie pointer to bufer descriptor to move.
+ *
+ * @return the physical memory address of the buffer. 
+ */
+static uintptr_t eth_allocate_rx_buf(void *iface, size_t buf_size, void **cookie)
```

Comment:
consider defining a dedicated type like `phys_addr_t` for physical addresses. If these are just normal buffers in the end, it could also be a simple `void*`